<center><h2>Individual Milestone: Bibliography & Resources (UCID 30229287)</h2></center><br>

<center><strong>Team 7:</strong> Analysis of Human Development Index (HDI)</center>


<b><u>Dataset</u></b><br>

<b>Table 1.</b> Human Development Index and its components<br>

<b>Dataset URL:</b> https://hdr.undp.org/data-center/human-development-index#/indicies/HDI<br>

<b>Dataset Description:</b> This dataset encompasses various facets of human development indices. It includes information on the Human Development Index (HDI) and its components, trends from 1990 to 2021, the Inequality-adjusted HDI, Gender Development Index, Gender Inequality Index, Multidimensional Poverty Index for developing countries, and the Planetary Pressures-adjusted HDI. These tables offer a comprehensive view of human development, covering socio-economic, gender-related, and environmental dimensions, providing a rich resource for analysis and exploration.<br>

<b>Dataset Importance:</b> This dataset is crucial for understanding human development through various indices like the Human Development Index (HDI), Gender Inequality Index, and Multidimensional Poverty Index. Covering trends from 1990 to 2021, it offers insights into development evolution over time, with a focus on equity and inclusivity. The inclusion of environmental considerations through the Planetary Pressures-adjusted HDI aligns with the current emphasis on sustainable development. This concise dataset serves as a vital resource for policymakers and researchers aiming to enhance global human development outcomes.
Queries


In [13]:
# Import the pandas library and alias it as 'pd'
import pandas as pd

# Read the data from the CSV file named "HDI_Table1.csv" into a DataFrame named 'cpl_locations'
cpl_locations = pd.read_csv("HDI_Table1.csv")

# Display the first 5 rows of the DataFrame to provide a quick overview of the loaded data
cpl_locations.head(5)

HDI rank                 Country  Human Development Index (HDI)   \
0         1             Switzerland                           0.962   
1         2                  Norway                           0.961   
2         3                 Iceland                           0.959   
3         4  Hong Kong, China (SAR)                           0.952   
4         5               Australia                           0.951   

   Life expectancy at birth  Expected years of schooling  \
0                   83.9872                    16.500299   
1                   83.2339                    18.185200   
2                   82.6782                    19.163059   
3                   85.4734                    17.278170   
4                   84.5265                    21.054590   

   Mean years of schooling  Gross national income (GNI) per capita  \
0                 13.85966                             66933.00454   
1                 13.00363                             64660.10622   
2                 13.76717                             55782.04981   
3                 12.22621                             62606.84540   
4                 12.72682                             49238.43335   

   GNI per capita rank minus HDI rank  HDI rank.1  
0                                   5           3  
1                                   6           1  
2                                  11           2  
3                                   6           4  
4                                  18           5

In [2]:
# Dictionary mapping old column names to new column names
new_column_names = {
    'HDI rank': 'HDIRank_2021',
    'Country': 'country',
    'Human Development Index (HDI) ':'hdi_value_index', 
    'Life expectancy at birth':'life_expectancy_at_birth', 
    'Expected years of schooling':'expected_years_of_schooling',
    'Mean years of schooling':'mean_years_of_schooling',
    'Gross national income (GNI) per capita': 'gni_per_capita',
    'GNI per capita rank minus HDI rank' : 'gni_per_capita_minus_rank',
    'HDI rank.1': 'HDIRank_2020'
}

# Use the rename method to rename the columns in the DataFrame 'cpl_locations'
# inplace=True modifies the DataFrame in place
cpl_locations.rename(columns=new_column_names, inplace=True)

# Display the first 5 rows of the DataFrame with the updated column names
cpl_locations.head(5)


HDIRank_2021                 country  hdi_value_index  \
0             1             Switzerland            0.962   
1             2                  Norway            0.961   
2             3                 Iceland            0.959   
3             4  Hong Kong, China (SAR)            0.952   
4             5               Australia            0.951   

   life_expectancy_at_birth  expected_years_of_schooling  \
0                   83.9872                    16.500299   
1                   83.2339                    18.185200   
2                   82.6782                    19.163059   
3                   85.4734                    17.278170   
4                   84.5265                    21.054590   

   mean_years_of_schooling  gni_per_capita  gni_per_capita_minus_rank  \
0                 13.85966     66933.00454                          5   
1                 13.00363     64660.10622                          6   
2                 13.76717     55782.04981                         11   
3                 12.22621     62606.84540                          6   
4                 12.72682     49238.43335                         18   

   HDIRank_2020  
0             3  
1             1  
2             2  
3             4  
4             5

In [3]:
# Import the mysql.connector library
import mysql.connector

# Import the errorcode module from mysql.connector
from mysql.connector import errorcode

# Establish a connection to the MySQL database
myconnection = mysql.connector.connect(
    user='prateek_kaushik',          # Username for database access
    password='YJBqdnSlqmk6F',        # Password for database access
    host='datasciencedb2.ucalgary.ca',  # Hostname or IP address of the MySQL server
    database='prateek_kaushik',      # Name of the database to connect to
    allow_local_infile=True          # Allow the use of local infile loading
)

# Display the connection object, showing details about the established connection
myconnection


In [4]:
# SQL query to create a table named 'hdi_index_components'
create_table_query = '''create table prateek_kaushik.hdi_index_components(
    HDIRank_2021 int NOT NULL,
    country varchar(40),
    hdi_index_components float, 
    life_expectancy_at_birth float,
    expected_years_of_schooling float,
    mean_years_of_schooling float,
    gni_per_capita float,
    gni_per_capita_minus_rank int,
    HDIRank_2020 int,
    created_at time,
    updated_at time
    );'''

# Create a cursor object for executing SQL queries
create_cursor = myconnection.cursor()

try:
    # Execute the CREATE TABLE query
    create_cursor.execute(create_table_query)
except mysql.connector.Error as err:
    # Handle specific error cases
    if err.errno == errorcode.ER_TABLE_EXISTS_ERROR:
        print('Oops! We already have that table!')
    else:
        print(err.msg)
else:
    # If the table creation is successful, print a success message
    print('Table Created Successfully!')

# Close the cursor to release resources
create_cursor.close()


Oops! We already have that table!


True

In [5]:
# Print the column names of the DataFrame as a list
# This line provides a quick overview of the columns present in the DataFrame
# It helps in understanding the structure of the dataset
print(cpl_locations.columns.tolist())

['HDIRank_2021', 'country', 'hdi_value_index', 'life_expectancy_at_birth', 'expected_years_of_schooling', 'mean_years_of_schooling', 'gni_per_capita', 'gni_per_capita_minus_rank', 'HDIRank_2020']


In [6]:
# SQL query to retrieve column information for the table 'hdi_index_components'
show_column_in_table = '''SHOW COLUMNS FROM prateek_kaushik.hdi_index_components;'''

# Create a cursor for executing SQL queries
show_cursor = myconnection.cursor()

try:
    # Execute the SHOW COLUMNS query
    show_cursor.execute(show_column_in_table)
    
    # Fetch all the rows (column information)
    columns = show_cursor.fetchall()
    
    # Print column names
    for column in columns:
        # Access the first element of the tuple (column name) and print it
        print(column[0])

except mysql.connector.Error as err:
    # Handle MySQL errors
    if err.errno == errorcode.ER_BAD_DB_ERROR:
        print('Error in query: Bad database')
    else:
        print(err.msg)

finally:
    # Close the cursor to release resources
    show_cursor.close()


HDIRank_2021
country
hdi_value_index
life_expectancy_at_birth
expected_years_of_schooling
mean_years_of_schooling
gni_per_capita
gni_per_capita_minus_rank
HDIRank_2020
created_at
updated_at


True

Before inserting the data into the SQL table, let us first check whether the `created_at` and `updated_at` are enabled for CURRENT_TIMESTAMP or not.

In [9]:
# SQL query to retrieve column information for 'created_at' and 'updated_at' columns in the table 'hdi_index_components'
verify_column_query = ''' SELECT COLUMN_NAME, COLUMN_DEFAULT 
                          FROM information_schema.COLUMNS
                          WHERE TABLE_SCHEMA = 'prateek_kaushik'
                          AND TABLE_NAME = 'hdi_index_components'
                          AND COLUMN_NAME IN ('created_at', 'updated_at');'''

# Create a cursor for executing SQL queries
verify_cursor = myconnection.cursor()

try:
    # Execute the SQL query to verify column information
    verify_cursor.execute(verify_column_query)

    # Fetch all the rows (column information)
    rows = verify_cursor.fetchall()

    # Print or process the fetched rows
    for row in rows:
        # Print the column name and its default value
        print(f"{row[0]} default value: {row[1]}")

except mysql.connector.Error as err:
    # Handle MySQL errors
    if err.errno == errorcode.ER_BAD_DB_ERROR:
        print('Error in query: Bad database')
    else:
        print(err.msg)

finally:
    # Close the cursor to release resources
    verify_cursor.close()


created_at default value: NULL
updated_at default value: NULL


The result indicates that the created_at and updated_at columns currently have NULL as their default values, which means they do not have explicit default values set.

Let us alter these columns to use the CURRENT_TIMESTAMP default value, we can use the ALTER TABLE statement.

In [10]:
# SQL query to alter the 'created_at' and 'updated_at' columns in the table 'hdi_index_components'
alter_table_query = '''ALTER TABLE prateek_kaushik.hdi_index_components 
                       MODIFY created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                       MODIFY updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP;'''

# Create a cursor for executing SQL queries
alter_cursor = myconnection.cursor()

try:
    # Execute the SQL query to alter the table
    alter_cursor.execute(alter_table_query)
except mysql.connector.Error as err:
    # Handle MySQL errors
    if err.errno == errorcode.ER_BAD_DB_ERROR:
        print('Error in query: Bad database')
    else:
        print(err.msg)
else:
    # If the ALTER TABLE query is successful, print a success message
    print('Table altered successfully.')

finally:
    # Close the cursor to release resources
    alter_cursor.close()


Table altered successfully.


True

Now let us check the types of `created_at` and `updated_at` whether they've been modified or not!

In [11]:
# SQL query to retrieve information about columns in the table 'hdi_index_components'
verify_column_query = '''SELECT COLUMN_NAME, COLUMN_DEFAULT
                         FROM information_schema.COLUMNS
                         WHERE TABLE_SCHEMA = 'prateek_kaushik'
                         AND TABLE_NAME = 'hdi_index_components'
                         AND COLUMN_NAME IN ('created_at', 'updated_at');'''

# Create a cursor for executing SQL queries
verify_cursor = myconnection.cursor()

try:
    # Execute the SQL query to retrieve column information
    verify_cursor.execute(verify_column_query)

    # Fetch all the rows
    rows = verify_cursor.fetchall()

    # Print or process the fetched rows
    for row in rows:
        # Print the column name and its default value
        print(f"{row[0]} default value: {row[1]}")

except mysql.connector.Error as err:
    # Handle MySQL errors
    if err.errno == errorcode.ER_BAD_DB_ERROR:
        print('Error in query: Bad database')
    else:
        print(err.msg)

finally:
    # Close the cursor to release resources
    verify_cursor.close()


created_at default value: current_timestamp()
updated_at default value: current_timestamp()


The alteration of the table has been successfully completed, enabling us to proceed with the insertion of data into the table.

In [12]:
# Create a cursor for executing SQL queries
insertCursor = myconnection.cursor()

# Convert column names to a string for SQL query
columnString = "`,`".join([str(currentColumn) for currentColumn in cpl_locations.columns.tolist()])

# Iterate through rows in the DataFrame and insert each row into the table
for i, currentRow in cpl_locations.iterrows():
    # Construct the SQL INSERT command dynamically
    insertCommand = "INSERT INTO `hdi_index_components` (`" + columnString + "`) VALUES (" + "%s," * (len(currentRow) - 1) + "%s)"
    
    # Execute the INSERT command with the values from the current row
    insertCursor.execute(insertCommand, tuple(currentRow))

# Commit the changes to the database
myconnection.commit()

# Close the cursor to release resources
insertCursor.close()


True

The code performs bulk insertion of data from a Pandas DataFrame (cpl_locations) into a MySQL table named hdi_index_components. It dynamically constructs and executes SQL INSERT commands for each row, facilitating efficient data transfer. The process is completed by committing the changes to the database, ensuring the successful integration of the dataset into the specified MySQL table.

In [13]:
# Define the SELECT query to retrieve the first 5 rows from the table
select_query = '''SELECT * FROM prateek_kaushik.hdi_index_components LIMIT 5;'''

# Create a cursor for executing SQL queries
select_cursor = myconnection.cursor()

try:
    # Execute the SELECT query
    select_cursor.execute(select_query)
    
    # Fetch all the rows from the result set
    rows = select_cursor.fetchall()
    
    # Print or process the fetched rows
    for row in rows:
        print(row)

except mysql.connector.Error as err:
    # Handle MySQL errors
    if err.errno == errorcode.ER_BAD_DB_ERROR:
        print('Error in query')
    else:
        print(err.msg)

finally:
    # Close the cursor to release resources
    select_cursor.close()


(1, 'Switzerland', 0.962, 83.9872, 16.5003, 13.8597, 66933.0, 5, 3, datetime.datetime(2023, 11, 20, 18, 54, 14), datetime.datetime(2023, 11, 20, 18, 54, 14))
(2, 'Norway', 0.961, 83.2339, 18.1852, 13.0036, 64660.1, 6, 1, datetime.datetime(2023, 11, 20, 18, 54, 14), datetime.datetime(2023, 11, 20, 18, 54, 14))
(3, 'Iceland', 0.959, 82.6782, 19.1631, 13.7672, 55782.1, 11, 2, datetime.datetime(2023, 11, 20, 18, 54, 14), datetime.datetime(2023, 11, 20, 18, 54, 14))
(4, 'Hong Kong, China (SAR)', 0.952, 85.4734, 17.2782, 12.2262, 62606.8, 6, 4, datetime.datetime(2023, 11, 20, 18, 54, 14), datetime.datetime(2023, 11, 20, 18, 54, 14))
(5, 'Australia', 0.951, 84.5265, 21.0546, 12.7268, 49238.4, 18, 5, datetime.datetime(2023, 11, 20, 18, 54, 14), datetime.datetime(2023, 11, 20, 18, 54, 14))


True

The provided code executes a SELECT query, retrieving the most recent five rows. The cursor fetches and prints each row, offering a verification step to confirm the successful insertion of records into the specified MySQL table. Additionally, the code reflects an optimization step undertaken before data insertion, involving an ALTER TABLE command to modify the 'created_at' and 'updated_at' columns. These alterations ensure that the columns are of the TIMESTAMP type with default values set to CURRENT_TIMESTAMP, enhancing data integrity and efficiency. Consequently, the SELECT query output provides a comprehensive view of the recently inserted data and the impact of the optimization on the 'created_at' and 'updated_at' columns.

In [17]:
# Define the SQL query to retrieve specific columns from the top 10 countries based on HDI ranking
top_10_countries = '''SELECT hdi.country, hdi.hdi_value_index, hdi.gni_per_capita
                      FROM hdi_index_components hdi
                      INNER JOIN hdi_index_components gni ON hdi.country = gni.country
                      WHERE hdi.HDIRank_2021 < 10;'''

# Create a cursor for executing SQL queries
top_10_cursor = myconnection.cursor()

try:
    # Execute the SQL query
    top_10_cursor.execute(top_10_countries)
    
    # Fetch all the rows from the result set
    rows = top_10_cursor.fetchall()
    
    # Print or process the fetched rows
    for row in rows:
        print(row)

except mysql.connector.Error as err:
    # Handle MySQL-specific errors
    if err.errno == errorcode.ER_BAD_DB_ERROR:
        print('Error in query')
    else:
        print(err.msg)

finally:
    # Close the cursor to release resources
    top_10_cursor.close()


('Switzerland', 0.962, 66933.0)
('Norway', 0.961, 64660.1)
('Iceland', 0.959, 55782.1)
('Hong Kong, China (SAR)', 0.952, 62606.8)
('Australia', 0.951, 49238.4)
('Denmark', 0.948, 60364.8)
('Sweden', 0.947, 54489.4)
('Ireland', 0.945, 76169.0)
('Germany', 0.942, 54534.2)


True

<b>Query Purpose:</b>
The above SQL query identifies and retrieves data on the top 10 countries with the highest Human Development Index (HDI) ranks in 2021.

<b>Summary:</b>
The result includes key information such as HDI values and Gross National Income (GNI) per capita for the top-ranking nations.

<b>Importance:</b>
This query offers a quick benchmark of leading countries in human development, aiding policymakers, development organizations, and analysts in strategic decision-making.

<b>Future Work & Analysis:</b>
The data can be leveraged for future analyses, including time-series evaluations, regional comparisons, and correlation studies to gain deeper insights into development trends.







In [18]:
# Define the SQL query to categorize countries based on their HDI values into 'Low', 'Medium', and 'High'
range_countries_hdi = '''SELECT CASE WHEN hdi_value_index < 0.6 THEN 'Low'
                                     WHEN hdi_value_index >= 0.6 AND hdi_value_index < 0.8 THEN 'Medium'
                                     WHEN hdi_value_index >= 0.8 THEN 'High'
                                END AS HDI_Range,
                                COUNT(*) AS CountryCount
                         FROM hdi_index_components
                         GROUP BY HDI_Range;'''

# Create a cursor for executing SQL queries
range_countries_hdi_cursor = myconnection.cursor()

try:
    # Execute the SQL query
    range_countries_hdi_cursor.execute(range_countries_hdi)

    # Fetch all the rows from the result set
    rows = range_countries_hdi_cursor.fetchall()

    # Print or process the fetched rows
    for row in rows:
        print(row)

except mysql.connector.Error as err:
    # Handle MySQL-specific errors
    if err.errno == errorcode.ER_BAD_DB_ERROR:
        print('Error in query')
    else:
        print(err.msg)

finally:
    # Close the cursor to release resources
    range_countries_hdi_cursor.close()


('High', 66)
('Low', 48)
('Medium', 77)


True

<b>Query Purpose:</b>
This SQL query categorizes countries by their Human Development Index (HDI) into Low, Medium, and High ranges, providing a count of countries in each category.

<b>Summary:</b>
The result offers a quick overview of country distribution based on development levels.

<b>Importance:</b> Policy Insight: Aids in tailoring development strategies for countries in specific HDI categories.
Progress Tracking: Provides a metric for monitoring overall development progress.

<b>Future Work & Analysis:</b>

Temporal Trends: Observing shifts in country distribution over time.</br>
Regional Dynamics: Exploring variations in HDI distribution among different regions.</br>
Correlation Studies: Investigating connections between HDI and other socio-economic indicators.

This analysis serves as a foundation for future projects, supporting in-depth studies on country dynamics and informing targeted development policies.

In [11]:
# SQL query to calculate average life expectancy for different HDI ranges
avg_life_expectancy = '''SELECT CASE WHEN hdi_value_index < 0.6 THEN 'Low'
                                     WHEN hdi_value_index >= 0.6 AND hdi_value_index < 0.8 THEN 'Medium'
                                     WHEN hdi_value_index >= 0.8 THEN 'High'
                                END AS HDI_Range,
                                AVG(life_expectancy_at_birth) AS AvgLifeExpectancy
                         FROM hdi_index_components
                         GROUP BY HDI_Range
                         ORDER BY AvgLifeExpectancy DESC;'''

# Create a cursor for executing SQL queries
avg_life_cursor = myconnection.cursor()

try:
    # Execute the average life expectancy query
    avg_life_cursor.execute(avg_life_expectancy)
    
    # Fetch all the rows from the query result
    rows = avg_life_cursor.fetchall()
    
    # Print or process the fetched rows
    for row in rows:
        print(row)

except mysql.connector.Error as err:
    # Handle MySQL errors
    if err.errno == errorcode.ER_BAD_DB_ERROR:
        print('Error in query')
    else:
        print(err.msg)

finally:
    # Close the cursor to release resources
    avg_life_cursor.close()


('High', 78.73505424730706)
('Medium', 70.86027957866719)
('Low', 61.83335828781128)


True

<b>Query Purpose:</b>
The SQL query calculates the average life expectancy at birth for countries categorized into Low, Medium, and High Human Development Index (HDI) ranges. The purpose is to analyze how life expectancy varies across different levels of human development.<br>

<b>Summary:</b>
The result presents the average life expectancy for countries in each HDI range:<br>

<ul>
    <li><b>High HDI Range:</b> 78.74 years</li>
    <li><b>Medium HDI Range:</b> 70.86 years</li>
    <li><o><b>Low HDI Range:</b> 61.83 years</li>
</ul>

<b>Importance:</b>

<b>Development Assessment:</b> Provides a holistic view of the relationship between HDI and life expectancy, aiding in assessing the overall development and well-being of countries.<br>

<b>Policy Implications:</b> Helps policymakers understand the impact of human development on life expectancy, guiding the formulation of targeted health and development policies.<br>

<b>Future Work & Analysis:</b><br>

<b>Temporal Analysis:</b> Extend the study over multiple years to observe trends in average life expectancy within each HDI range over time.<br>

<b>Correlation Studies:</b> Investigate correlations between life expectancy, HDI components, and other socio-economic indicators for a more comprehensive understanding of development dynamics.<br>

<b>Healthcare Strategies:</b> Use the insights to inform strategies for improving healthcare systems in countries with lower life expectancies.<br>

This analysis serves as a foundational element for future research, contributing valuable insights into the intricate interplay between human development and life expectancy.

<b><u>Citations</u></b>

[1] U. Nations, “Human Development Index,” United Nations. Available: https://hdr.undp.org/data-center/human-development-index. [Accessed: Nov. 06, 2023]

[2] U. Nations, “Terms of use,” United Nations. Available: https://hdr.undp.org/terms-use. [Accessed: Nov. 06, 2023]

[3] B. Herre, P. Arriagada, and M. Roser, “The Human Development Index and related indices: what they are and what we can learn from them,” Our World in Data, Nov. 2023, Available: https://ourworldindata.org/human-development-index. [Accessed: Nov. 06, 2023]

[4] “What Is the Human Development Index (HDI)?,” Investopedia. Available: https://www.investopedia.com/terms/h/human-development-index-hdi.asp. [Accessed: Nov. 06, 2023]

[5] G. Dongre, “What is Human Development Index?,” Medium, Jun. 12, 2021. Available: https://ganeshdongre.medium.com/what-is-human-development-index-9f2f991dfd88. [Accessed: Nov. 06, 2023]

[6] Python mysql—Join. (2020, March 16). GeeksforGeeks. https://www.geeksforgeeks.org/python-mysql-join/
